File: src/ui/banner.ts
// ASCII banner with a diagonal violet → pink → cyan truecolor gradient.

import { colorEnabled } from "./colors.ts";

const ART = [
  " ███╗   ███╗ █████╗ ███╗   ██╗ ██████╗  █████╗ ",
  " ████╗ ████║██╔══██╗████╗  ██║██╔════╝ ██╔══██╗",
  " ██╔████╔██║███████║██╔██╗ ██║██║  ███╗███████║",
  " ██║╚██╔╝██║██╔══██║██║╚██╗██║██║   ██║██╔══██║",
  " ██║ ╚═╝ ██║██║  ██║██║ ╚████║╚██████╔╝██║  ██║",
  " ╚═╝     ╚═╝╚═╝  ╚═╝╚═╝  ╚═══╝ ╚═════╝ ╚═╝  ╚═╝",
  "          ██████╗██╗     ██╗                   ",
  "         ██╔════╝██║     ██║                    ",
  "         ██║     ██║     ██║                    ",
  "         ██║     ██║     ██║                    ",
  "         ╚██████╗███████╗██║                    ",
  "          ╚═════╝╚══════╝╚═╝                    ",
];

// Three gradient stops, swept diagonally (top-left → bottom-right).
const STOPS: Array<[number, number, number]> = [
  [149, 76, 233], // violet
  [244, 114, 182], // pink
  [34, 211, 238], // cyan
];

function lerp(a: number, b: number, t: number): number {
  return Math.round(a + (b - a) * t);
}

/** Sample the 3-stop gradient at t ∈ [0,1]. */
function gradient(t: number): [number, number, number] {
  const seg = t < 0.5 ? 0 : 1;
  const local = (t - seg * 0.5) * 2;
  const [a, b] = [STOPS[seg], STOPS[seg + 1]];
  return [lerp(a[0], b[0], local), lerp(a[1], b[1], local), lerp(a[2], b[2], local)];
}

export function banner(version?: string): string {
  const H = ART.length;
  const W = Math.max(...ART.map((l) => l.length));
  // Diagonal sweep: rows count ~2.2 columns each so the angle looks natural
  // with terminal cells being taller than they are wide.
  const span = W + 2.2 * (H - 1);

  const lines = ART.map((line, y) => {
    if (!colorEnabled) return line;
    let out = "";
    let last = "";
    for (let x = 0; x < line.length; x++) {
      const chr = line[x];
      if (chr === " ") {
        out += chr;
        continue;
      }
      const [r, g, b] = gradient(Math.min(1, (x + 2.2 * y) / span));
      const code = `\x1b[38;2;${r};${g};${b}m`;
      if (code !== last) {
        out += code;
        last = code;
      }
      out += chr;
    }
    return out + "\x1b[0m";
  });

  const text = `terminal manga reader 🎲${version ? ` · v${version}` : ""}`;
  const pad = " ".repeat(Math.max(0, Math.floor((W - text.length) / 2)));
  const tagline = colorEnabled ? `${pad}\x1b[2m${text}\x1b[0m` : pad + text;
  return "\n" + lines.join("\n") + "\n" + tagline + "\n";
}

/** Banner is shown only on an interactive TTY, unless explicitly disabled. */
export function shouldShowBanner(configShow: boolean): boolean {
  if (process.env.MANGA_CLI_NO_BANNER === "1") return false;
  if (!process.stdout.isTTY) return false;
  return configShow;
}

File: go-rewrite/cmd/manga-cli/banner.go
// The ASCII banner with a diagonal violet → pink → cyan gradient.

package main

import (
	"fmt"
	"os"
	"strings"

	"github.com/devcheckra1n/manga-cli/go-rewrite/internal/ui"
)

var bannerArt = []string{
	" ███╗   ███╗ █████╗ ███╗   ██╗ ██████╗  █████╗ ",
	" ████╗ ████║██╔══██╗████╗  ██║██╔════╝ ██╔══██╗",
	" ██╔████╔██║███████║██╔██╗ ██║██║  ███╗███████║",
	" ██║╚██╔╝██║██╔══██║██║╚██╗██║██║   ██║██╔══██║",
	" ██║ ╚═╝ ██║██║  ██║██║ ╚████║╚██████╔╝██║  ██║",
	" ╚═╝     ╚═╝╚═╝  ╚═╝╚═╝  ╚═══╝ ╚═════╝ ╚═╝  ╚═╝",
	"          ██████╗██╗     ██╗                   ",
	"         ██╔════╝██║     ██║                    ",
	"         ██║     ██║     ██║                    ",
	"         ██║     ██║     ██║                    ",
	"         ╚██████╗███████╗██║                    ",
	"          ╚═════╝╚══════╝╚═╝                    ",
}

var bannerStops = [3][3]int{{149, 76, 233}, {244, 114, 182}, {34, 211, 238}}

func bannerGradient(t float64) (int, int, int) {
	seg := 0
	if t >= 0.5 {
		seg = 1
	}
	local := (t - float64(seg)*0.5) * 2
	a, b := bannerStops[seg], bannerStops[seg+1]
	lerp := func(x, y int) int { return x + int(float64(y-x)*local) }
	return lerp(a[0], b[0]), lerp(a[1], b[1]), lerp(a[2], b[2])
}

func banner(version string) string {
	h := len(bannerArt)
	w := 0
	for _, l := range bannerArt {
		w = max(w, len([]rune(l)))
	}
	span := float64(w) + 2.2*float64(h-1)
	var out strings.Builder
	out.WriteString("\n")
	for y, line := range bannerArt {
		if !ui.ColorEnabled {
			out.WriteString(line + "\n")
			continue
		}
		last := ""
		for x, chr := range []rune(line) {
			if chr == ' ' {
				out.WriteRune(' ')
				continue
			}
			t := min(1.0, (float64(x)+2.2*float64(y))/span)
			r, g, b := bannerGradient(t)
			code := fmt.Sprintf("\x1b[38;2;%d;%d;%dm", r, g, b)
			if code != last {
				out.WriteString(code)
				last = code
			}
			out.WriteRune(chr)
		}
		out.WriteString("\x1b[0m\n")
	}
	text := "terminal manga reader 🎲"
	if version != "" {
		text += " · v" + version
	}
	pad := strings.Repeat(" ", max(0, (w-len([]rune(text)))/2))
	out.WriteString(pad + ui.Dim(text) + "\n")
	return out.String()
}

func shouldShowBanner(configShow bool) bool {
	if os.Getenv("MANGA_CLI_NO_BANNER") == "1" {
		return false
	}
	return configShow && ui.IsTTY()
}
